nixos/modules/services/web-apps/firefly-iii.nix

```diff
@@ -0,0 +1,356 @@
+    AmbientCapabilities = "";
+    ProtectSystem = "strict";
+    ProtectKernelTunables = true;
+    ProtectKernelModules = true;
+    ProtectControlGroups = true;
+    ProtectClock = true;
+    ProtectHostname = true;
+    ProtectHome = "tmpfs";
+    ProtectKernelLogs = true;
+    ProtectProc = "invisible";
+    ProcSubset = "pid";
+    PrivateNetwork = false;
+    RestrictAddressFamilies = "AF_INET AF_INET6 AF_UNIX";
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [
+      "@system-service @resources"
+      "~@obsolete @privileged"
+    ];
+    RestrictSUIDSGID = true;
+    RemoveIPC = true;
+    NoNewPrivileges = true;
+    RestrictRealtime = true;
+    RestrictNamespaces = true;
+    LockPersonality = true;
+    PrivateUsers = true;
+  };
+
+in {
+
+  options.services.firefly-iii = {
+
+    enable = mkEnableOption "Firefly III: A free and open source personal finance manager";
+
+    dataDir = mkOption {
+      description = mdDoc "Firefly III data directory";
+      default = "/var/lib/firefly-iii";
+      type = path;
+    };
+
+    user = mkOption {
+      type = str;
+      default = defaultUser;
+      description = mdDoc "User account under which firefly-iii runs.";
+    };
+
+    group = mkOption {
+      type = str;
+      default = defaultGroup;
+      description = mdDoc ''
+        Group under which firefly-iii runs. It is best to set this to the group
+        of whatever webserver is being used as the frontend.
+      '';
+    };
+
+    package = mkOption {
+      type = package;
+      default = firefly-pkg;
+      defaultText = literalExpression "pkgs.firefly-iii";
+      description = mdDoc ''
+        The firefly-iii package served by php-fpm and the webserver of choice.
+        This option can be used to point the webserver to the correct root. It
+        may also be used to set the package to a different version, say a
+        development version.
+      '';
+    };
+
+    # PHP
+    poolConfig = mkOption {
+      type = attrsOf (oneOf [ str int bool ]);
+      default = {
+        "pm" = "dynamic";
+        "pm.max_children" = 32;
+        "pm.start_servers" = 2;
+        "pm.min_spare_servers" = 2;
+        "pm.max_spare_servers" = 4;
+        "pm.max_requests" = 500;
+      };
+      description = mdDoc ''
+        Options for the Firefly III PHP pool. See the documentation on <literal>php-fpm.conf</literal>
+        for details on configuration directives.
+      '';
+    };
+
+    settings = mkOption {
+      default = {};
+      description = mdDoc ''
+        Options for firefly-iii configuration. Refer to
+        <https://github.com/firefly-iii/firefly-iii/blob/main/.env.example> for
+        details on supported values. All <option>_FILE values supported by
+        upstream are supported here.
+      '';
+      example = literalExpression ''
+        {
+          APP_ENV = "production";
+          APP_URL = "https://example.com";
+          APP_KEY = "SomeKeyOfExactly32Chars";
```

Comment:
Can we store the APP_KEY securely?

Reply:
We can, we can use `APP_KEY_FILE` instead. I'll change the example to that.